Getting Started with OpenTelemetry and Datadog guide (#21439)
DataDog/documentation: content/en/getting_started/opentelemetry/_index.md

The Calendar sample application is already instrumented for you.

In `CalendarController.java`, the following code instruments the Calendar application using the OpenTelemetry API:

```
private String getDate() {
   Span span = GlobalOpenTelemetry.getTracer("calendar").spanBuilder("getDate").startSpan();
```

When the Calendar application runs, the `getDate()` call generates traces and spans. Next, you will configure the application to send this data to the Datadog backend.

### Configuring the OTLP Receiver 

The Calendar application sends data from the OpenTelemetry SDK to the [OpenTelemetry Protocol (OTLP) receiver][10] in the OpenTelemetry Collector. In the Collector configuration file---which lives at `./src/main/resources/otelcol-config.yaml`---the following code configures the OTLP receiver to receive metrics, traces, and logs:

```
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318
...
service:
  pipelines:
    traces:
      receivers: [otlp]
    metrics:
      receivers: [otlp]
    logs:
      receivers: [otlp]
```

### Configuring the Datadog Exporter

> The Datadog Exporter sends data collected by the OTLP receiver to the Datadog backend. In `otelcol-config.yaml`, the following configures the Datadog Exporter:

Reviewer: ```suggestion
The Datadog Exporter sends data collected by the OTLP Receiver to the Datadog backend.

1. Go to the `otelcol-config.yaml` file.
2. The following lines configure the Datadog Exporter to send observability data to Datadog:
```

```
service:
  pipelines:
    metrics:
      exporters: [datadog]
    traces:
      exporters: [datadog]
    logs:
      exporters: [datadog]
```

This code suffices to send runtime metrics, traces, and logs to Datadog. However, [*infrastructure metrics* require additional changes][4]. 

### Configuring your application to send infrastructure metrics

As an example, you will configure your OpenTelemetry Collector to forward Docker container metrics. 

> *Reminder: You must use Linux to send infrastructure metrics from the OpenTelemetry Collector to Datadog.*

Reviewer: ```suggestion
<div class="alert alert-info">To send infrastructure metrics from the OpenTelemetry Collector to Datadog, you must use Linux.</div>
```

To collect container metrics, configure the [Docker stats receiver][5] in your Datadog Exporter: 

1. Add a `docker_stats` block to the `receivers` section of `otel-config.yaml`:

```
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        Endpoint: 0.0.0.0:4318
  # add the following block
  docker_stats: 
    endpoint: unix:///var/run/docker.sock # default; if this is not the Docker socket path, update to the correct path
    metrics:
      container.network.io.usage.rx_packets:
        enabled: true
      container.network.io.usage.tx_packets:
        enabled: true
      container.cpu.usage.system:
        enabled: true
      container.memory.rss:
        enabled: true
      container.blockio.io_serviced_recursive:
        enabled: true
      container.uptime:
        enabled: true
      container.memory.hierarchical_memory_limit:
        enabled: true
```

2. Update `service.pipelines.metrics.receivers` to include `docker_stats`:

```
service:
  pipelines:
    metrics:
      receivers: [otlp, docker_stats] # <- update this line!
```

> Your application now sends container metrics to Datadog for viewing in Datadog.

Reviewer: ```suggestion
This configuration allows the Calendar application to send container metrics to Datadog for you to explore in Datadog.
```

### Using unified tagging to correlate observability data

[Unified service tagging][6] ties observability data together in Datadog so you can navigate across metrics, traces, and logs with consistent tags.

> In the Calendar application's Docker Compose file, `./deploys/docker/docker-compose-otel.yml`, the following lines are already included to enable correlation between application traces and other observability data: 

Reviewer: ```suggestion
The Calendar application is already configured with unified service tagging:

1. Go to the Calendar application's Docker Compose file at `./deploys/docker/docker-compose-otel.yml`.
2. The following lines enable the correlation between application traces and other observability data: 
```

Reviewer: Side note: I noticed there's mixed file naming of the YAML file extensions. If this reflects how it looks in the sample app that's fine, but if not, we should keep them consistently `.yaml` or `.yml`.

Author: This is how it is in the sample app... docker-compose-otel uses ".yml" and otelcol-config uses ".yaml"

```
environment:
  - OTEL_SERVICE_NAME=calendar-otel
  - OTEL_LOGS_EXPORTER=otlp
  - OTEL_RESOURCE_ATTRIBUTES=deployment.environment=docker,host.name=otelcol-docker
```

## Running the application to generate observability data

> Run the Calendar application with the OpenTelemetry SDK to generate and forward observability data to Datadog:  

Reviewer: ```suggestion
All of the initial setup and configuration is complete. To start generating and forwarding observability data to Datadog, you need to run the Calendar application with the OpenTelemetry SDK:
```

1. Run the application from the `calendar/` folder:
   ```
   docker compose -f deploys/docker/docker-compose-otel.yml up
   ```

   This command creates a Docker container with the OpenTelemetry Collector and the Calendar service.

> 2. Test the Calendar application by running the following from another terminal window:

Reviewer: ```suggestion
2. To test that the Calendar application is running correctly, execute the following command from another terminal window:
```

   ```
   curl localhost:9090/calendar 
   ```

>    You should get a response like:

Reviewer: ```suggestion
3. Verify that you receive a response like:
```

   ```
   {“date":"2022-12-30"}
   ```

> With each call, the Calendar application metrics, traces, and logs, which are forwarded to the OpenTelemetry Collector, the Datadog Exporter, and finally the Datadog backend. 

Reviewer: ```suggestion
Each call to the Calendar application results in metrics, traces, and logs being forwarded to the OpenTelemetry Collector, then to the Datadog Exporter, and finally to the Datadog backend. 
```

> Note: The Calendar application uses the probabilistic sampler processor, so only 30% of traces sent through the application reach the target backend. Run the curl command several times to ensure at least one trace exports to the Datadog backend. 

Reviewer: ```suggestion
**Note**: The Calendar application uses the probabilistic sampler processor, so only 30% of traces sent through the application reach the target backend. Run the `curl` command several times to ensure at least one trace exports to the Datadog backend. 
```

## Exploring observability data in Datadog

Now, you can use Datadog to explore the Calendar application's observability data.   

*Note: It may take a few minutes for your trace data to appear in Datadog.*

### Exploring Runtime and Infrastructure Metrics

View runtime and infrastructure metrics to visualize, monitor, and measure the performance of your applications, hosts, containers, and processes.

> 1. Hover over **APM** in the left-side panel and select **Service Catalog**.
> 2. Hover over the `calendar-otel` service and select **Full Page**.
> 3. Scroll to the bottom panel and select:
> 

Reviewer: ```suggestion
1. Go to **APM** > **Service Catalog**.
2. Hover over the `calendar-otel` service and select **Full Page**.
3. Scroll to the bottom panel and select:

```

   * **Infrastructure Metrics** to see your Docker container metrics, such as CPU and memory usage.
   * **JVM Metrics** to see runtime metrics, such as heap usage and thread count. 

   TODO: insert screenshot

### Exploring Logs

View logs to monitor and troubleshoot application and system operations.

Select **Logs** in the left-side panel to explore logs generated by the Calendar application. 

TODO: insert screenshot

### Exploring Traces

View traces and spans to observe the status and performance of requests processed by your application.

> 1. Hover over **APM** in the left-side panel and select **Traces**.

Reviewer: ```suggestion
1. Go to **APM** > **Traces**.
```

> 2. Find the **Service** section in the filter menu on the left-hand side, and select the `calendar-otel` service to list all `calendar-otel` traces:

Reviewer: ```suggestion
2. Find the **Service** section in the filter menu, and select the `calendar-otel` facet to display all `calendar-otel` traces:
```

   TODO: insert screenshot

3. [Explore your `calendar-otel` traces][8]. 

   To start, click on a trace to open the trace side panel and find more details about the trace and its spans. For example, the [Flame Graph][9] captures how much time was spent in each component of the Calendar execution path:

   TODO: insert screenshot

4. Notice that you can select **Infrastructure**, **Metrics**, or **Logs** in the bottom panel to correlate your trace with other observability data. 

   TODO: insert screenshot

>    You can also take the inverse approach: for example, in the **Logs** section of Datadog, you can click on a log and correlate it back to the respective trace.

Reviewer: ```suggestion
```
Not sure this is necessary to include. I think one way is sufficient, but let me know if you disagree.

   TODO: insert screenshot

## Further Reading

{{< partial name="whats-next/whats-next.html" >}}

[1]: https://www.datadoghq.com/free-datadog-trial/
[2]: https://app.datadoghq.com/organization-settings/api-keys/
[3]: https://docs.docker.com/compose/install/

> [4]: https://docs.datadoghq.com/opentelemetry/collector_exporter/

Reviewer: ```suggestion
[4]: /opentelemetry/collector_exporter/
```

[5]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/dockerstatsreceiver/

> [6]: https://docs.datadoghq.com/getting_started/tagging/unified_service_tagging/

Reviewer: ```suggestion
[6]: /getting_started/tagging/unified_service_tagging/
```

[7]: https://app.datadoghq.com/services

> [8]: https://docs.datadoghq.com/tracing/glossary/#trace

Reviewer: ```suggestion
[8]: /tracing/glossary/#trace
```

[9]: https://www.datadoghq.com/knowledge-center/distributed-tracing/flame-graph/

> [10]: https://docs.datadoghq.com/opentelemetry/collector_exporter/otlp_receiver/

Reviewer: ```suggestion
[10]: /opentelemetry/collector_exporter/otlp_receiver/
```

### Configuring the OTLP Receiver 

The Calendar application is already configured to send data from the OpenTelemetry SDK to the [OpenTelemetry Protocol (OTLP) receiver][10] in the OpenTelemetry Collector.

1. Go to the Collector configuration file located at: `./src/main/resources/otelcol-config.yaml`
2.  The following lines configure the OTLP Receiver to receive metrics, traces, and logs:  

    ```yaml
    receivers:
      otlp:
        protocols:
          grpc:
            endpoint: 0.0.0.0:4317
          http:
            endpoint: 0.0.0.0:4318
    ...
    service:
      pipelines:
        traces:
          receivers: [otlp]
        metrics:
          receivers: [otlp]
        logs:
          receivers: [otlp]
    ```

### Configuring the Datadog Exporter

The Datadog Exporter sends data collected by the OTLP Receiver to the Datadog backend.

1. Go to the `otelcol-config.yaml` file.
2. The following lines configure the Datadog Exporter to send observability data to Datadog:  

    ```yaml
    service:
      pipelines:
        metrics:
          exporters: [datadog]
        traces:
          exporters: [datadog]
        logs:

>           exporters: [datadog]

Reviewer 2: Yes.

    ```

This configuration allows the Datadog Exporter to send runtime metrics, traces, and logs to Datadog. However, sending infrastructure metrics requires additional configuration.

### Configuring your application to send infrastructure metrics

In this example, configure your OpenTelemetry Collector to forward Docker container metrics.

<div class="alert alert-info">To send infrastructure metrics from the OpenTelemetry Collector to Datadog, you must use Linux.</div>

To collect container metrics, configure the [Docker stats receiver][5] in your Datadog Exporter: 

> 1. Add a `docker_stats` block to the `receivers` section of `otel-config.yaml`:

Reviewer: We have [a note above this](https://docs-staging.datadoghq.com/dominique.alessi/otelgs/getting_started/opentelemetry/#configuring-your-application-to-send-infrastructure-metrics): "To send infrastructure metrics from the OpenTelemetry Collector to Datadog, you must use Linux."

Do you think we need to add a mention that this is a limitation of Docker stats receiver?

> 1. Add a `docker_stats` block to the `receivers` section of `otel-config.yaml`:

Reviewer 2: Yes, we should mention that. Re-emphasis that it works with Linux is good idea.

            enabled: true
          container.memory.rss:
            enabled: true
          container.blockio.io_serviced_recursive:
            enabled: true
          container.uptime:
            enabled: true
          container.memory.hierarchical_memory_limit:
            enabled: true
    ```

2. Update `service.pipelines.metrics.receivers` to include `docker_stats`:

    ```yaml
    service:
      pipelines:
        metrics:
          receivers: [otlp, docker_stats] # <- update this line!
    ```

This configuration allows the Calendar application to send container metrics to Datadog for you to explore in Datadog.

### Using unified tagging to correlate observability data

[Unified service tagging][6] ties observability data together in Datadog so you can navigate across metrics, traces, and logs with consistent tags.

The Calendar application is already configured with unified service tagging:

1. Go to the Calendar application's Docker Compose file at `./deploys/docker/docker-compose-otel.yml`.
2. The following lines enable the correlation between application traces and other observability data: 

    ```yaml
    environment:
      - OTEL_SERVICE_NAME=calendar-otel

>       - OTEL_LOGS_EXPORTER=otlp

Reviewer: Maybe we add this to a new subsection like this?

### Configuring your application to send logs with OTLP

The Calendar application uses the OpenTelemetry logging exporter in its Logback configuration to send logs with OpenTelemetry Layer Processor (OTLP).

1. Go to the Calendar application's Logback XML configuration file at `/src/main/resources/logback.xml`.
2. The following lines define the `OpenTelemetry` appender:

   ```xml
    <appender name="OpenTelemetry" class="io.opentelemetry.instrumentation.logback.appender.v1_0.OpenTelemetryAppender">
        <immediateFlush>true</immediateFlush>
        <captureExperimentalAttributes>true</captureExperimentalAttributes>
        <captureKeyValuePairAttributes>true</captureKeyValuePairAttributes>
      </appender>
   ```

Additionally, an environment variable configures the OpenTelemetry environment to export logs:

1. Go to the Calendar application's Docker Compose file at `./deploys/docker/docker-compose-otel.yml`.
2. The `OTEL_LOGS_EXPORTER=otlp` configuration allows the logs to be sent with OTLP.

>       - OTEL_LOGS_EXPORTER=otlp

Reviewer 2: Yup , sounds good. After defining the appender we need to use it as well. https://github.com/DataDog/opentelemetry-examples/blob/bbb938cad421bbe492141f9bf7e3f89e475f4596/apps/rest-services/java/calendar/src/main/resources/logback.xml#L18

>       - OTEL_LOGS_EXPORTER=otlp

Reviewer: Got it! I'll add this section then and also include adding it as you show on L18. 🙏

1. Run the application from the `calendar/` folder:
   ```sh
   docker compose -f deploys/docker/docker-compose-otel.yml up
   ```

   This command creates a Docker container with the OpenTelemetry Collector and the Calendar service.

2. To test that the Calendar application is running correctly, execute the following command from another terminal window:
   ```sh
   curl localhost:9090/calendar 
   ```

3. Verify that you receive a response like:
   ```
   {“date":"2022-12-30"}
   ```

Each call to the Calendar application results in metrics, traces, and logs being forwarded to the OpenTelemetry Collector, then to the Datadog Exporter, and finally to the Datadog backend. 

**Note**: The Calendar application uses the probabilistic sampler processor, so only 30% of traces sent through the application reach the target backend. Run the `curl` command several times to ensure at least one trace exports to the Datadog backend. 

## Exploring observability data in Datadog

Now, you can use Datadog to explore the Calendar application's observability data.   

**Note**: It may take a few minutes for your trace data to appear in Datadog.

> ### Exploring runtime and infrastructure metrics

Reviewer 3: It would be great to also add a section to illustrate infra host list view for the workflow of exploring Infrastructure level health in this section.

Reviewer: So you are thinking we need to expand the Infrastructure Metrics note into a section that also shows what specifically is shown in there (e.g., CPU Usage by Pod)?
